pkg/retry/retry.go

```diff
@@ -77,3 +114,86 @@ func OnError(
 		}
 	}
 }
+
+// Call executes a function repeatedly, according to the retry strategy until
+// it succeeds or returns an `ErrNonRetryable`, which indicates that no more
+// retries should be attempted.
+//
+// If no retry strategy is provided, it defaults to an exponential backoff strategy
+// with 25 max retries, 500ms initial delay, and 30000ms max delay.
+//
+// Returns the result from the work function and any error that occurred if retries are exhausted.
+func Call[T any](
+	ctx context.Context,
+	work func() (T, error),
+	retryStrategy ...RetryStrategyFunc,
+) (T, error) {
+	var (
+		result T
+		err    error
+	)
+	if retryStrategy == nil {
+		retryStrategy = []RetryStrategyFunc{DefaultExponentialDelay}
+	}
+
+	for retryCount := 0; ; retryCount++ {
+		result, err = work()
+
+		// Stop retrying and return the result if no error occurred
+		if err == nil {
+			return result, err
+		}
+
+		// Stop retrying and return the result if the error is non-retryable
+		if ErrNonRetryable.Is(err) {
+			return result, err
+		}
+
+		// Stop retrying and return if the error is a non-transient gRPC error
+		status, isGRPCError := status.FromError(err)
+		if isGRPCError && !slices.Contains(transientGRPCErrorCodes, status.Code()) {
+			return result, err
+		}
+
+		if !retryStrategy[0](ctx, retryCount) {
```

Comment:
#PUC on the approach/logic here.

- What's with the [0]?
- What is it doing?
- Is there an associated TODO?